src/main/java/graphql/validation/RulesVisitor.java

```diff
@@ -88,139 +79,108 @@ public void enter(Node node, List<Node> ancestors) {
     }
 
-    private void checkFragmentSpread(FragmentSpread fragmentSpread, List<AbstractRule> rules, List<Node> ancestors) {
-        for (AbstractRule rule : rules) {
-            rule.checkFragmentSpread(fragmentSpread);
-        }
-        List<AbstractRule> rulesVisitingFragmentSpreads = getRulesVisitingFragmentSpreads(rules);
+    private void checkFragmentSpread(FragmentSpread node, List<AbstractRule> rules, List<Node> ancestors) {
+        rules.forEach(r -> r.checkFragmentSpread(node));
+
+        ImmutableList<AbstractRule> rulesVisitingFragmentSpreads = filterRulesVisitingFragmentSpreads(rules, true);
         if (rulesVisitingFragmentSpreads.size() > 0) {
-            FragmentDefinition fragment = validationContext.getFragment(fragmentSpread.getName());
-            if (fragment != null && !ancestors.contains(fragment)) {
-                new LanguageTraversal(ancestors).traverse(fragment, new RulesVisitor(validationContext, rulesVisitingFragmentSpreads, true));
+            FragmentDefinition fragment = validationContext.getFragment(node.getName());
+            if (fragment != null && !ancestors.contains(fragment) && !visitedFragmentSpreads.contains(node.getName())) {
+                // Manually traverse into the FragmentDefinition
+                visitedFragmentSpreads.add(node.getName());
+                rulesStack.push(rulesVisitingFragmentSpreads);
+                fragmentSpreadVisitDepth++;
+                new LanguageTraversal(ancestors).traverse(fragment, this);
+                fragmentSpreadVisitDepth--;
+                rulesStack.pop();
             }
         }
     }
 
-    private List<AbstractRule> getRulesVisitingFragmentSpreads(List<AbstractRule> rules) {
-        List<AbstractRule> result = new ArrayList<>();
-        for (AbstractRule rule : rules) {
-            if (rule.isVisitFragmentSpreads()) result.add(rule);
-        }
-        return result;
-    }
-
+    private void checkFragmentDefinition(FragmentDefinition node, ImmutableList<AbstractRule> rules) {
+        ImmutableList<AbstractRule> scopeRules = rules;
 
-    private void checkFragmentDefinition(FragmentDefinition fragmentDefinition, List<AbstractRule> rules) {
-        if (!subVisitor) {
-            rulesToSkipByUntilNode.put(fragmentDefinition, new ArrayList<>(rulesVisitingFragmentSpreads));
-            rulesToSkip.addAll(rulesVisitingFragmentSpreads);
-        }
-
-
-        for (AbstractRule rule : rules) {
-            if (!subVisitor && (rule.isVisitFragmentSpreads())) continue;
-            rule.checkFragmentDefinition(fragmentDefinition);
+        // If we've encountered a FragmentDefinition and we got here without coming through a
+        // FragmentSpread, then suspend all isVisitFragmentSpread rules for this subtree.
+        // Expect these rules to be checked when when the FragmentSpread is traversed
+        if (fragmentSpreadVisitDepth == 0) {
+            scopeRules = filterRulesVisitingFragmentSpreads(rules, false);
```

Comment:
so these are basically all rules which don't want to visit fragment spreads, meaning all Rules except 3.